### ISSUE TITLE
[BUG] Parquet Reader crashes when reading ZSTD compressed Parquet on Pascal

### ISSUE THREAD

**Issue by mythrocks** [rapidsai/cudf#11834]
When CUDF Parquet reader reads (certain) ZSTD compressed Parquet files on Pascal GPUs, it encounters `cudaErrorIllegalAddress`, causing the process to `terminate()`.

This bug was discovered in `spark-rapids` validation, when running tests on Pascal GPUs. (Relevant bug: https://github.com/NVIDIA/spark-rapids/issues/6614). Further details are available in [this comment](https://github.com/NVIDIA/spark-rapids/issues/6614#issuecomment-1262856672). Reading ZSTD compressed Parquet via `spark-rapids` on Pascal hardware leads to the `spark` application crashing.

Reproduction of the failure on other architectures (Turing, Ampere) has remained elusive. One only sees this failure on Pascal.

The failure may be reproduced with the [attached synthetically generated (zipped) Parquet file](https://github.com/NVIDIA/spark-rapids/files/9679294/borked.zstd.parquet.zip). (Note: This needs to be unzipped to be read.)
The following gtest reproduces the failure:
```c++
TEST_F(ParquetReaderTest, Borked_read_of_ZSTD)
{
  auto read_opts = cudf_io::parquet_reader_options::builder(
                     cudf::io::source_info{"/tmp/borked.zstd.parquet"});
  auto const result = cudf::io::read_parquet(read_opts);
  cudf::test::print(result.tbl->get_column(0));
}
```
The test results in the following:
```
$ gtests/PARQUET_TEST --gtest_filter=ParquetReaderTest.Bork\*   
                                                              
Note: Google Test filter = ParquetReaderTest.Bork*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from ParquetReaderTest
[ RUN      ] ParquetReaderTest.Borked_read_of_ZSTD
terminate called after throwing an instance of 'thrust::system::system_error'
  what():  cudaErrorIllegalAddress: an illegal memory access was encountered
fish: Job 1, 'gtests/PARQUET_TEST --gtest_fil…' terminated by signal SIGABRT (Abort)
```

Note that the attached Parquet file is readable via `parquet-tools`, etc.:
```
$ parquet-tools show /tmp/borked.zstd.parquet | head -10                                                                                            
+---------------+---------------+
|             a |             b |
|---------------+---------------|
|   9.22337e+18 |   7.07165e+18 |
|  -3.63696e+18 | nan           |
|  -3.32315e+18 |   1           |
|   5.68865e+18 |  -1.63744e+17 |
|  -8.3441e+18  |   6.18778e+18 |
|  -8.98935e+18 |   9.22337e+18 |
|  -6.41698e+17 |  -4.3473e+18  |
```

Note: Not only does `compute-sanitizer --tool memcheck/racecheck` not catch the error, it also allows the read to succeed. Possible Heisenbug. 

One is inclined to believe that this has to do with ZSTD (de)compression, since https://github.com/NVIDIA/spark-rapids/issues/6614 indicates that ZSTD ORC might also be having similar trouble (incorrect reads, not necessarily crashy ones).

**Comment by mythrocks** [comment]
Apologies. It would have been prudent to include details about the card on which the read fails:
```
$ nvidia-smi -L                                                                                                                                         
GPU 0: NVIDIA TITAN X (Pascal) (UUID: GPU-XXXXXXXX-XXXX-...)
```

**Comment by mythrocks** [comment]
Sorry I missed this before: There seems to be a race-condition in the Parquet reader, maybe?
```
========= Warning: Race reported between Write access at 0xb670 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *)
=========     and Read access at 0xb310 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [4 hazards]
=========     and Read access at 0xb3b8 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [4 hazards]
=========     and Read access at 0xb410 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [20 hazards]
=========     and Read access at 0xb468 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [8 hazards]
=========     and Read access at 0xb4d0 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [8 hazards]
=========     and Read access at 0xb528 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [8 hazards]
========= 
========= Warning: Race reported between Write access at 0x2fa28 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *)
=========     and Read access at 0x2f6e8 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [4 hazards]
=========     and Read access at 0x2f858 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [4 hazards]
=========     and Read access at 0x2f8d0 in zstd::decompression_kernel(const unsigned char *const *, unsigned char *const *, unsigned long *, const unsigned long *, unsigned long, unsigned char *, unsigned long *, unsigned long, zstd::ZstdFrame *, zstd::DeviceBlockShare *, int *, int *, int *, int *) [12 hazards]
========= 
========= Error: Race reported between Read access at 0x3428 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long)
=========     and Write access at 0xdd10 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long) [768 hazards]
========= 
========= Warning: Race reported between Read access at 0xc828 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long)
=========     and Write access at 0xd770 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long) [136 hazards]
========= 
========= Warning: Race reported between Read access at 0xc7f8 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long)
=========     and Write access at 0xd768 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long) [68 hazards]
========= 
========= Warning: Race reported between Read access at 0xc810 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long)
=========     and Write access at 0xd748 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long) [68 hazards]
========= 
========= Warning: Race reported between Read access at 0xc818 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long)
=========     and Write access at 0xd758 in cudf::io::parquet::gpu::gpuDecodePageData(cudf::io::parquet::gpu::PageInfo *, cudf::device_span<const cudf::io::parquet::gpu::ColumnChunkDesc, (unsigned long)18446744073709551615>, unsigned long, unsigned long) [68 hazards]
========= 
```
Notice: That's one `Error` on `cudf::io::parquet::gpu::gpuDecodePageData()`, and several warnings on `zstd::decompression_kernel()`.
cc @vuule , who knows this better than I would.

**Comment by vuule** [comment]
CC @nvdbaranec for the Parquet reader race

**Comment by jbrennan333** [comment]
The bug was in nvcomp-2.4, fixed in nvcomp-2.4.1.
This was fixed by this rapids-cmake PR: https://github.com/rapidsai/rapids-cmake/pull/286

**Comment by mythrocks** [comment]
This is addressed in `nvcomp-2.4.1`. Plus, CUDF now disables ZSTD reads on Pascal unless it's using `nvcomp-2.4.1`: https://github.com/rapidsai/cudf/pull/11856
